broker/src/main/java/io/zeebe/broker/clustering/atomix/storage/snapshot/DbSnapshotStoreFactory.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.0. You may not use this file
+ * except in compliance with the Zeebe Community License 1.0.
+ */
+package io.zeebe.broker.clustering.atomix.storage.snapshot;
+
+import io.atomix.protocols.raft.storage.snapshot.SnapshotStore;
+import io.atomix.protocols.raft.storage.snapshot.SnapshotStoreFactory;
+import io.zeebe.util.ZbLogger;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.NavigableMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+import org.agrona.IoUtil;
+import org.slf4j.Logger;
+
+/**
+ * Loads existing snapshots in memory, cleaning out old and/or invalid snapshots if present.
+ *
+ * <p>The current load strategy is to lookup all files directly under the {@code
+ * SNAPSHOTS_DIRECTORY}, try to extract {@link DbSnapshotMetadata} from them, and if not possible
+ * skip them (and print out a warning).
+ *
+ * <p>The metadata extraction is done by parsing the directory name using '%d-%d-%d-%d', where in
+ * order we expect: index, term, timestamp, and position.
+ */
+public class DbSnapshotStoreFactory implements SnapshotStoreFactory {
+  static final String SNAPSHOTS_DIRECTORY = "snapshots";
+  static final String PENDING_DIRECTORY = "pending";
+  private static final Logger LOGGER = new ZbLogger(DbSnapshotStoreFactory.class);
+
+  @Override
+  public SnapshotStore createSnapshotStore(final Path root, final String prefix) {
+    final var snapshots = new ConcurrentSkipListMap<DbSnapshotId, DbSnapshot>();
```

Comment:
Why we need here a concurrent map?

Reply:
Same as above, because the snapshot store is accessed by multiple threads, and different threads will put snapshots in the store. The other option is to always do I/O to check the latest snapshots, which I'd hoped to avoid.